pipeline_dp/dataframes.py

```diff
@@ -0,0 +1,189 @@
+
+from dataclasses import dataclass
+from typing import Any, Dict, Iterable, List, Optional
+
+import pipeline_dp
+
+
+@dataclass
+class Columns:
+    privacy_key: str
+    partition_key: str
+    value: Optional[str]
+
+
+@dataclass
+class ContributionBounds:
+    max_partitions_contributed: Optional[int] = None
+    max_contributions_per_partition: Optional[int] = None
+    min_value: Optional[float] = None
+    max_value: Optional[float] = None
+
+
+@dataclass
+class Budget:
+    epsilon: float
+    delta: float = 0
+
+    def __post_init__(self):
+        if self.epsilon <= 0:
+            raise ValueError(f"epsilon must be positive, not {self.epsilon}.")
+        if self.delta < 0:
+            raise ValueError(f"delta must be non-negative, not {self.delta}.")
+
+
+class Query:
+    """DP Query on DataFrames.
+
+    Warning: Don't create it directly, use QueryBuilder().
+    """
+
+    def __init__(self, df, columns: Columns, metrics: Dict[pipeline_dp.Metric,
+                                                           List[str]],
+                 contribution_bounds: ContributionBounds, public_keys):
+        self._df = df
+        self._columns = columns
+        self._metrics = metrics
+        self._contribution_bounds = contribution_bounds
+        self._public_partitions = public_keys
+
+    def run_query(self,
+                  budget: Budget,
+                  noise_kind: Optional[pipeline_dp.NoiseKind] = None):
+        raise NotImplementedError("Running query is not yet implemented")
+
+
+class QueryBuilder:
+    """Builds DP queries on (Spark, Pandas, Beam) DataFrames.
+
+    It uses a Builder pattern, each public function except of build_query
+    returns the reference to itself.
+    The usage is
+    QueryBuilder().groupby(...).aggregation1(...)....aggregation_n().build_query()
+
+    Example:
+        # The following query computes restaurant visits and total money
+        # spent:
+        query = QueryBuilder(df, "user_id")
+        .groupby("day", max_groups_contributed=3, max_contributions_per_group=1)
+        .count()
+        .sum("money_spent", min_value=0, max_value=100)
+        .build_query()
+    """
+
+    def __init__(self, df, privacy_unit_column: str):
+        """Constructor.
+
+        Args:
+            df: (Spark, Pandas, Beam) DataFrame with data to anonymize.
+            privacy_key_column: column from `df` with privacy key.
+        """
+        if privacy_unit_column not in df.columns:
+            raise ValueError(
+                f"Column {privacy_unit_column} is not present in DataFrame")
+        self._df = df
+        self._privacy_unit_column = privacy_unit_column
+        self._groupby_column = None
+        self._value_column = None
+        self._metrics = {}  # map from pipeline_dp.Metric -> output column name
+        self._contribution_bounds = ContributionBounds()
+        self._public_keys = None
+
+    def groupby(self,
+                column: str,
```

Comment:
how will the support for grouping by more columns be supported? Will we need to call it once again?

Reply:
ah, I see, it can be called only once.

But maybe then make it a list? so that we have correct interface from the beginning. I guess the same then applies to privacy unit. But up to you if it is difficult to make right now.